File: app/src/main/java/com/example/sakura/felicareader/felicahistory/SuicaPasmoHistory.java
package com.example.sakura.felicareader.felicahistory;

/**
 * Suica,Pasmo履歴コード
 * 参考：https://osdn.jp/projects/felicalib/wiki/suica
 */

public class SuicaPasmoHistory {

    public int remain;

    public SuicaPasmoHistory(){
    }

    public static SuicaPasmoHistory parse(byte[] res, int off) {
        SuicaPasmoHistory self = new SuicaPasmoHistory();
        self.init(res, off);
        return self;
    }

    private void init(byte[] res, int off) {
        this.remain  = toInt(res, off, 11,10); //11-10: Suica,Pasmo残高
    }

    private int toInt(byte[] res, int off, int... idx) {
        int num = 0;
        for (int i=0; i<idx.length; i++) {
            num = num << 8;
            num += ((int)res[off+idx[i]]) & 0x0ff;
        }
        return num;
    }

    public String toString() {
        String str = "残高："+remain+"円";
        return str;
    }
}

File: app/src/main/java/sakura/felica/felicahistory/SuicaPasmoHistory.kt
package sakura.felica.felicahistory

/**
 * Suica,Pasmo履歴コード
 * 参考：https://osdn.jp/projects/felicalib/wiki/suica
 */

class SuicaPasmoHistory {

  private var remain: Int = 0

  private fun init(res: ByteArray, off: Int) {
    this.remain = toInt(res, off, 11, 10) //11-10: Suica,Pasmo残高
  }

  private fun toInt(res: ByteArray, off: Int, vararg idx: Int): Int {
    var num = 0
    idx.map {
      num = num shl 8
      num += res[off + it].toInt() and 0x0ff
    }

    return num
  }

  override fun toString(): String = "残高：" + remain + "円"

  companion object {

    fun parse(res: ByteArray, off: Int): SuicaPasmoHistory {
      val self = SuicaPasmoHistory()
      self.init(res, off)
      return self
    }
  }
}
